delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Projection.scala

```diff
@@ -57,35 +59,61 @@ final class Projection private[stream] (
-        progressRef <- Ref[Task].of(progress)
-        stream    = projection.streamF
-          .apply(progress.offset)(status)(signal)
-          .interruptWhen(signal)
-          .onFinalizeCaseWeak {
-            case ExitCase.Error(th) => status.update(_.failed(th))
-            case ExitCase.Completed => Task.unit // streams stopped through a signal still finish as Completed
-            case ExitCase.Canceled  => Task.unit // the status is updated by the logic that cancels the stream
-          }.mapAccumulate(progress) {
-          case (acc, msg) if msg.offset.value > progress.offset.value => (acc + msg, msg)
-          case (acc, msg)                                  => (acc, msg)
-        }.groupWithin(batch.maxElements, batch.maxInterval).evalTap { elements =>
-          elements.last.fold(Task.unit) { case (newProgress, _) =>
-            progressRef.set(newProgress) >> saveProgress(newProgress)
-          }
-        }
-          .compile
-          .drain
-        // update status to Running at the beginning and to Completed at the end if it's still running
-        task      = status.update(_ => ExecutionStatus.Running) >> stream >> status.update(s => if (s.isRunning) ExecutionStatus.Completed else s)
-        fiber    <- task.start
-        fiberRef <- Ref[Task].of(fiber)
-      } yield new Projection(projection.metadata.name, status, progressRef, signal, fiberRef)
+
+  private val logger: Logger = Logger[Projection]
+
+  def apply(
+      projection: CompiledProjection,
+      fetchProgress: UIO[Option[ProjectionProgress]],
+      saveProgress: ProjectionProgress => UIO[Unit],
+      saveFailedElems: List[FailedElem] => UIO[Unit]
+  )(implicit batch: BatchConfig): Task[Projection] =
+    for {
+      status      <- Ref[Task].of[ExecutionStatus](ExecutionStatus.Pending)
+      signal      <- SignallingRef[Task, Boolean](false)
+      progress    <- fetchProgress.map(_.getOrElse(ProjectionProgress.NoProgress))
+      progressRef <- Ref[Task].of(progress)
+      stream       = projection.streamF
+                       .apply(progress.offset)(status)(signal)
+                       .interruptWhen(signal)
+                       .onFinalizeCaseWeak {
+                         case ExitCase.Error(th) => status.update(_.failed(th))
+                         case ExitCase.Completed => Task.unit // streams stopped through a signal still finish as Completed
+                         case ExitCase.Canceled  => Task.unit // the status is updated by the logic that cancels the stream
+                       }
+                       .mapAccumulate(progress) {
+                         case (acc, msg) if msg.offset.value > progress.offset.value => (acc + msg, msg)
+                         case (acc, msg)                                             => (acc, msg)
+                       }
+                       .groupWithin(batch.maxElements, batch.maxInterval)
+                       .map { elements =>
```

Comment:
* Join the `map` and the `evalTap` in a single `evalTap`
* Extract the content in a function outside the for comprehension